image_view/src/image_saver.cpp

```diff
@@ -0,0 +1,284 @@
+  }
+
+  bool callbackStartSave(
+    const std::shared_ptr<rmw_request_id_t> request_header,
+    const std::shared_ptr<std_srvs::srv::Trigger::Request> request,
+    std::shared_ptr<std_srvs::srv::Trigger::Response> response)
+  {
+    (void)request_header;
+    (void)request;
+    RCLCPP_INFO(this->get_logger(), "Received start saving request");
+    start_time_ = this->now();
+    end_time_ = rclcpp::Time(0);
+
+    response->success = true;
+    return true;
+  }
+
+  bool callbackEndSave(
+    const std::shared_ptr<rmw_request_id_t> request_header,
+    const std::shared_ptr<std_srvs::srv::Trigger::Request> request,
+    std::shared_ptr<std_srvs::srv::Trigger::Response> response)
+  {
+    (void)request_header;
+    (void)request;
+    RCLCPP_INFO(this->get_logger(), "Received end saving request");
+    end_time_ = this->now();
+
+    response->success = true;
+    return true;
+  }
+
+  void callbackWithoutCameraInfo(const sensor_msgs::msg::Image::ConstSharedPtr & image_msg)
+  {
+    if (is_first_image_) {
+      is_first_image_ = false;
+
+      // Wait a tiny bit to see whether callbackWithCameraInfo is called
+      rclcpp::sleep_for(std::chrono::milliseconds(1));
+    }
+
+    if (has_camera_info_) {
+      return;
+    }
+
+    // saving flag priority:
+    //  1. request by service.
+    //  2. request by topic about start and end.
+    //  3. flag 'save_all_image'.
+    if (!save_image_service && request_start_end) {
+      if (start_time_ == rclcpp::Time(0)) {
+        return;
+      } else if (start_time_ > image_msg->header.stamp) {
+        return;  // wait for message which comes after start_time
+      } else if ((end_time_ != rclcpp::Time(0)) && (end_time_ < image_msg->header.stamp)) {
+        return;  // skip message which comes after end_time
+      }
+    }
+
+    // save the image
+    std::string filename;
+    if (!saveImage(image_msg, filename)) {
+      return;
+    }
+
+    count_++;
+  }
+
+  void callbackWithCameraInfo(
+    const sensor_msgs::msg::Image::ConstSharedPtr & image_msg,
+    const sensor_msgs::msg::CameraInfo::ConstSharedPtr & info)
+  {
+    has_camera_info_ = true;
+
+    if (!save_image_service && request_start_end) {
+      if (start_time_ == rclcpp::Time(0)) {
+        return;
+      } else if (start_time_ > image_msg->header.stamp) {
+        return;  // wait for message which comes after start_time
+      } else if ((end_time_ != rclcpp::Time(0)) && (end_time_ < image_msg->header.stamp)) {
+        return;  // skip message which comes after end_time
+      }
+    }
+
+    // save the image
+    std::string filename;
+    if (!saveImage(image_msg, filename)) {
+      return;
+    }
+
+    // save the CameraInfo
+    if (info) {
+      filename = filename.replace(filename.rfind("."), filename.length(), ".ini");
+      camera_calibration_parsers::writeCalibration(filename, "camera", *info);
+    }
+
+    count_++;
+  }
+};
+
+}  // namespace image_view
+
+int main(int argc, char** argv)
```

Comment:
Main in new file and register this as a component

Reply:
Would you prefer I do this for all of the stand-alone tools? I considered it but I didn't think they would be that useful in a pipeline (though I could see this one).

Reply:
I'd say "sure, doesn't hurt".

Reply:
All nodes are now components and have a separate single-node executable.